openpype/plugins/publish/integrate_new.py

```diff
@@ -110,801 +156,494 @@ class IntegrateAssetNew(pyblish.api.InstancePlugin):
-                instance,
-                self.integrated_file_sizes)
-
-            self.log.debug("__ representation: {}".format(representation))
-            destination_list.append(dst)
-            self.log.debug("__ destination_list: {}".format(destination_list))
-            instance.data['destination_list'] = destination_list
-            representations.append(representation)
-            published_representations[repre_id] = {
-                "representation": representation,
-                "anatomy_data": template_data,
-                "published_files": published_files
-            }
-            self.log.debug("__ representations: {}".format(representations))
-
-        # Remove old representations if there are any (before insertion of new)
-        if existing_repres:
-            repre_ids_to_remove = []
-            for repre in existing_repres:
-                repre_ids_to_remove.append(repre["_id"])
-            io.delete_many({"_id": {"$in": repre_ids_to_remove}})
-
-        for rep in instance.data["representations"]:
-            self.log.debug("__ rep: {}".format(rep))
-
-        io.insert_many(representations)
-        instance.data["published_representations"] = (
-            published_representations
+            # todo: reduce/simplify what is returned from this function
+            prepared = self.prepare_representation(repre,
+                                                   template_name,
+                                                   existing_repres_by_name,
+                                                   version,
+                                                   instance_stagingdir,
+                                                   instance)
+
+            for src, dst in prepared["transfers"]:
+                # todo: add support for hardlink transfers
+                file_transactions.add(src, dst)
+
+            prepared_representations.append(prepared)
+
+        if not prepared_representations:
+            # Even though we check `instance.data["representations"]` earlier
+            # this could still happen if all representations were tagged with
+            # "delete" and thus are skipped for integration
+            raise RuntimeError("No representations prepared to publish.")
+
+        # Each instance can also have pre-defined transfers not explicitly
+        # part of a representation - like texture resources used by a
+        # .ma representation. Those destination paths are pre-defined, etc.
+        # todo: should we move or simplify this logic?
+        for src, dst in instance.data.get("transfers", []):
+            file_transactions.add(src, dst, mode=FileTransaction.MODE_COPY)
+        for src, dst in instance.data.get("hardlinks", []):
+            file_transactions.add(src, dst, mode=FileTransaction.MODE_HARDLINK)
+
+        # Bulk write to the database
+        # todo: Can we move this even to after the file transfers?
```

Comment:
I believe we could

Reply:
It would mean however that if two machines would publish at the same time there's a higher chance they publish into the same version without warning and both write a unique version id with the same version number. Especially if files copying takes long that chance increases. It's not entirely safe. Unless we'd maybe do another check JUST before writing to the database and after the file transfer. We could e.g. do that only if more time passed than "safety amount".

Reply:
Damn. you're totally right and that's actually why it's where it is, to reserve the version. Not sure there any benefit of moving it later really apart from visual clarity in the code (io operation clustered together). probably not worth the hassle

Reply:
Updated the comment to clarify why we write version in advance. Only reason I had to move it later is to ensure file transfer went OK and try and implement the database writing into a single transaction that can be rolled back easily. But I'll consider it ok for now. 👍